File: backend/app/schemas/catalog.py
"""Pydantic schemas for the /api/catalog surface.

These mirror the geny-executor introspection dataclasses. We stay close to
the library's `to_dict` output so the backend is a thin pass-through —
callers don't have to re-implement the shape on the frontend.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class ArtifactInfoResponse(BaseModel):
    """One artifact's metadata as returned by the catalog."""

    stage: str
    name: str
    description: str = ""
    version: str = "1.0"
    stability: str = "stable"
    requires: List[str] = Field(default_factory=list)
    is_default: bool = False
    provides_stage: bool = True
    extra: Dict[str, Any] = Field(default_factory=dict)


class SlotIntrospectionResponse(BaseModel):
    slot_name: str
    description: str = ""
    required: bool = False
    current_impl: str = ""
    available_impls: List[str] = Field(default_factory=list)
    impl_schemas: Dict[str, Optional[Dict[str, Any]]] = Field(default_factory=dict)
    impl_descriptions: Dict[str, str] = Field(default_factory=dict)


class ChainIntrospectionResponse(BaseModel):
    chain_name: str
    description: str = ""
    current_impls: List[str] = Field(default_factory=list)
    available_impls: List[str] = Field(default_factory=list)
    impl_schemas: Dict[str, Optional[Dict[str, Any]]] = Field(default_factory=dict)
    impl_descriptions: Dict[str, str] = Field(default_factory=dict)


class StageIntrospectionResponse(BaseModel):
    """Full introspection of a single stage+artifact combination."""

    stage: str
    artifact: str
    order: int
    name: str
    category: str = ""
    artifact_info: ArtifactInfoResponse
    config_schema: Optional[Dict[str, Any]] = None
    config: Dict[str, Any] = Field(default_factory=dict)
    strategy_slots: Dict[str, SlotIntrospectionResponse] = Field(default_factory=dict)
    strategy_chains: Dict[str, ChainIntrospectionResponse] = Field(default_factory=dict)
    tool_binding_supported: bool = True
    model_override_supported: bool = True
    extra: Dict[str, Any] = Field(default_factory=dict)


class StageSummaryResponse(BaseModel):
    """Lightweight stage summary used by the /catalog/stages list route."""

    order: int
    module: str  # canonical module name, e.g. "s06_api"
    name: str  # short name, e.g. "api"
    category: str = ""
    default_artifact: str = "default"
    artifact_count: int = 0


class StageListResponse(BaseModel):
    stages: List[StageSummaryResponse]


class ArtifactListResponse(BaseModel):
    stage: str
    artifacts: List[ArtifactInfoResponse]


class FullCatalogResponse(BaseModel):
    """Bootstrap response for the Environment Builder UI."""

    stages: List[StageIntrospectionResponse]

File: frontend/src/types/catalog.ts
/* Types mirroring the session-less /api/catalog responses.
 *
 * The backend returns the geny_executor library's introspection dicts as-is,
 * so these shapes are the library's canonical output. Field names MUST match
 * the library's `StageIntrospection.to_dict()` output; when the two drift the
 * Builder silently renders blank tabs.
 *
 * Two schema shapes live here:
 *   - JsonSchemaObject  — the library's native output (JSON Schema)
 *   - ConfigSchema       — the flat Record<field, FieldInfo> that
 *                          ConfigSchemaForm consumes. Convert with
 *                          `flattenJsonSchema` before rendering.
 */

// ── Flat ConfigSchema (consumed by ConfigSchemaForm) ────

export type ConfigFieldType =
  | "string"
  | "integer"
  | "number"
  | "boolean"
  | "enum"
  | "object"
  | "array"
  | "any";

export interface ConfigFieldInfo {
  name: string;
  type: ConfigFieldType;
  required?: boolean;
  default?: unknown;
  description?: string;
  /** Present when type === "enum". */
  choices?: string[];
  /** Nested object schema. */
  properties?: Record<string, ConfigFieldInfo>;
  /** Element schema when type === "array". */
  items?: ConfigFieldInfo;
  /** Optional numeric bounds (type === "integer" | "number"). */
  minimum?: number;
  maximum?: number;
}

export type ConfigSchema = Record<string, ConfigFieldInfo>;

// ── JSON Schema shape emitted by the library ────────────

/** One property entry inside a JSON Schema `properties` map. */
export interface JsonSchemaField {
  type?: string;
  title?: string;
  description?: string;
  default?: unknown;
  enum?: unknown[];
  enumLabels?: Record<string, string>;
  minimum?: number;
  maximum?: number;
  minLength?: number;
  maxLength?: number;
  pattern?: string;
  items?: JsonSchemaField;
  properties?: Record<string, JsonSchemaField>;
  required?: string[];
  "x-ui-widget"?: string;
  [key: string]: unknown;
}

/** Top-level JSON Schema object — library's `config_schema` output. */
export interface JsonSchemaObject {
  type: "object";
  title?: string;
  properties: Record<string, JsonSchemaField>;
  required?: string[];
  [key: string]: unknown;
}

// ── Artifact / introspection shapes (library-native) ────

export interface ArtifactInfo {
  stage: string;
  name: string;
  description: string;
  version: string;
  stability: string;
  requires: string[];
  is_default: boolean;
  provides_stage: boolean;
  extra: Record<string, unknown>;
}

export interface SlotIntrospection {
  slot_name: string;
  description: string;
  required: boolean;
  current_impl: string;
  available_impls: string[];
  /** Map of impl_name → that impl's JSON Schema (or null). */
  impl_schemas: Record<string, JsonSchemaObject | null>;
  impl_descriptions: Record<string, string>;
}

export interface ChainIntrospection {
  chain_name: string;
  description: string;
  current_impls: string[];
  available_impls: string[];
  impl_schemas: Record<string, JsonSchemaObject | null>;
  impl_descriptions: Record<string, string>;
}

export interface StageIntrospection {
  stage: string;
  artifact: string;
  order: number;
  name: string;
  category: string;
  artifact_info: ArtifactInfo;
  config_schema: JsonSchemaObject | null;
  /** Default stage-level config for this artifact. */
  config: Record<string, unknown>;
  strategy_slots: Record<string, SlotIntrospection>;
  strategy_chains: Record<string, ChainIntrospection>;
  tool_binding_supported: boolean;
  model_override_supported: boolean;
  /** True for stages the pipeline can't run without (Input / API / Parse /
   * Yield today). UIs must force `active=true` on these. Added with
   * geny-executor >=0.13.3 — treated as `false` for older backends. */
  required?: boolean;
  extra?: Record<string, unknown>;
}

export interface StageSummary {
  order: number;
  module: string;
  name: string;
  category: string;
  default_artifact: string;
  artifact_count: number;
}

export interface CatalogFullResponse {
  stages: StageIntrospection[];
}

export interface CatalogStageListResponse {
  stages: StageSummary[];
}

export interface CatalogArtifactListResponse {
  stage: string;
  artifacts: ArtifactInfo[];
}
